source/fuzz/transformation_replace_params_with_struct.cpp

```diff
@@ -0,0 +1,296 @@
+        function) {
+      return false;
+    }
+  }
+
+  // We already know that the function has at least |parameter_id.size()|
+  // parameters.
+
+  // Check that all parameters have supported types.
+  if (!std::all_of(parameter_id.begin(), parameter_id.end(),
+                   [ir_context](uint32_t id) {
+                     const auto* type = ir_context->get_type_mgr()->GetType(
+                         fuzzerutil::GetTypeId(ir_context, id));
+                     return type && IsParameterTypeSupported(*type);
+                   })) {
+    return false;
+  }
+
+  // Check that a relevant OpTypeStruct exists in the module.
+  std::vector<uint32_t> component_type_ids;
+  for (auto id : message_.parameter_id()) {
+    component_type_ids.push_back(fuzzerutil::GetTypeId(ir_context, id));
+  }
+
+  if (!fuzzerutil::MaybeGetStructType(ir_context, component_type_ids)) {
+    return false;
+  }
+
+  // Check that |callee_id_to_fresh_composite_id| is valid.
+  for (const auto* inst :
+       fuzzerutil::GetCallers(ir_context, function->result_id())) {
+    // Check that the callee is present in the map. It's ok if the map contains
+    // more ids that there are callees (those ids will not be used).
+    if (!message_.caller_id_to_fresh_composite_id().contains(
+            inst->result_id())) {
+      return false;
+    }
+  }
+
+  // Check that all fresh ids are unique and fresh.
+  std::vector<uint32_t> fresh_ids = {message_.fresh_function_type_id(),
+                                     message_.fresh_parameter_id()};
+
+  for (const auto& entry : message_.caller_id_to_fresh_composite_id()) {
+    fresh_ids.push_back(entry.second);
+  }
+
+  return !fuzzerutil::HasDuplicates(fresh_ids) &&
+         std::all_of(fresh_ids.begin(), fresh_ids.end(),
+                     [ir_context](uint32_t id) {
+                       return fuzzerutil::IsFreshId(ir_context, id);
+                     });
+}
+
+void TransformationReplaceParamsWithStruct::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto* function = fuzzerutil::GetFunctionFromParameterId(
+      ir_context, message_.parameter_id(0));
+  assert(function);
+
+  // Create a new struct type.
+  std::vector<uint32_t> struct_components_ids;
+  for (auto id : message_.parameter_id()) {
+    struct_components_ids.push_back(fuzzerutil::GetTypeId(ir_context, id));
+  }
+
+  auto struct_type_id =
+      fuzzerutil::MaybeGetStructType(ir_context, struct_components_ids);
+  assert(struct_type_id &&
+         "IsApplicable should've guaranteed that this value isn't equal to 0");
+
+  // Add new parameter to the function.
+  function->AddParameter(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpFunctionParameter, struct_type_id,
+      message_.fresh_parameter_id(), opt::Instruction::OperandList()));
+
+  fuzzerutil::UpdateModuleIdBound(ir_context, message_.fresh_parameter_id());
+
+  // Compute indices of parameters.
+  std::vector<uint32_t> param_indices;
```

Comment:
This variable is super-important - it gets used loads from now on - so I suggest giving it a slightly more descriptive name, like: "indices_of_removed_params", and changing the comment to say that we compute indices of the parameters that will be removed, so that we can work out which actual parameters we need to use to populate a struct instance at every call site.

Reply:
Fixed in 877e4d4.